### PR TITLE
3 registration and authorization

### PR DESCRIPTION
**Pull request by Nooruzbai**
Added Google authentication

### CODE REVIEW

File: src/project/routers/auth.py [hunk 1]
```diff
@@ -94,8 +102,69 @@ async def get_authenticated_page(
...
+    return response
+
+#Google Authentication
+def oauth_user(
+    email: str,
+    oauth_provider: OauthProvider,
+    session: Session,
+):
+    user = User.get_by_email(session=session, email=email)
+    if not user:
```
Comment (Agahan14) [inline src/project/routers/auth.py:119]: It doesn't make sense

File: src/project/utils/errors/errors.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from fastapi import HTTPException
```
Comment (Agahan14) [inline src/project/utils/errors/errors.py:1]: Custom exceptions do not benefit the project in any way

File: src/project/routers/auth.py [hunk 3]
```diff
@@ -94,8 +102,69 @@ async def get_authenticated_page(
...
+    except Exception as e:
+        raise UserDataError(status_code=400, detail=str(e))
+    google_user_info = token.get("userinfo")
+    if not google_user_info:
+        raise UserDataError(status_code=303, detail=OAUTH_NO_USER_INFO)
+    email = google_user_info.get("email")
+    if not email:
+        raise UserDataError(status_code=303, detail=OAUTH_NO_EMAIL)
+    try:
+        user = oauth_user(
```
[the remark this answers is not in the document]
Comment (Nooruzbai) [reply src/project/routers/auth.py:157]: Ok, will add a registration

File: src/project/routers/auth.py [hunk 4]
```diff
@@ -9,46 +9,59 @@
...
         db.commit()
         db.refresh(new_user)
+        return RedirectResponse(url="/login", status_code=303)
     except IntegrityError:
         db.rollback()
-        return HTMLResponse(content="Email already registered.", status_code=400)
-    return RedirectResponse(url="/auth/login", status_code=303)
+        return HTMLResponse(content=EMAIL_ALREADY_REGISTERED, status_code=400)
+
+    return RedirectResponse(url="/login", status_code=303)
```
Comment (Agahan14) [inline src/project/routers/auth.py:64]: on line 73 you are already returning redirect

File: src/project/routers/auth.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import os
```
Comment (Agahan14) [inline src/project/routers/auth.py:1]: you also need to format the whole file with ruff, I don't see you formatting

File: src/project/routers/auth.py [hunk 6]
```diff
@@ -9,46 +9,59 @@
...
 @router.get("/register", tags=["auth"], response_class=HTMLResponse)
-async def get_register_user(request: Request) -> HTMLResponse:
+async def register_user(request: Request) -> HTMLResponse:
     return templates.TemplateResponse("register.html", {"request": request})
 
 
 #Post view for user registration
 @router.post("/register", tags=["auth"], response_class=HTMLResponse)
-async def post_register_user(
+async def register_user(
```
Comment (Agahan14) [inline src/project/routers/auth.py:37]: I remember I asked you to rename the methods, but it’s better to leave it as you did before. Rename them as get_register, post_register

**Pull request merged**